Source/DafnyCore/Compilers/Java/Compiler-java.cs

```diff
@@ -945,6 +933,51 @@ protected class ClassWriter : IClassWriter {
       return new ClassWriter(this, wRestOfBody, wCtorBody);
     }
 
+    /// <summary>
+    /// For each type parameter X in "typeParametersForClass" that needs a type descriptor,
+    ///   * Write "protected TypeDescriptor<X> _td_X;" to wTypeFields
+    ///     -- each entry is terminated by a newline
+    ///   * Write "TypeDescriptor<X> _td_X" to wCtorParams
+    ///     -- entries are separated by a comma
+    ///   * Write "_td_X" to wCallArguments
+    ///     -- entries are separated by a comma
+    ///   * Write "this._td_X := _td_X;" to wCtorBody
+    ///     -- each entry is terminated by a newline
+    /// Any of the writer parameters may be null, so long as at least one is non-null.
+    /// The method returns the number type descriptors written.
+    /// </summary>
+    int EmitTypeDescriptorsForClass(List<TypeParameter> typeParametersForClass, TopLevelDecl cls,
```

Comment:
What's the difference between this code and the C# version?

Reply:
There is indeed a LOT of near-duplication of code throughout the compilers. But I find it out of scope for this PR to fix that.

Reply:
Even when introducing more of that near duplication? Are you proposing some future larger change that would resolve multiple duplication problems at once?